Ceci est un fichier Jupyter Notebook.
Ce fichier contient mes démarches et explications au devoir.
Je, Hossein Hajmirbaba (300208487), ai effectué ce devoir seul.
Je suis concient que c'est possible que lorsque vous allez run le program vous allez obtenir des résultats différent. C'est quelques choses que je comprends toujours pas comment je peux faire. Si vous pourriez écrire dans les commentaires de correction pour comment la prochaine fois je peux analyser mes résultats.
Merci.

# 1. Explication du problème

L'objectif principal de ce devoir est d'effectué une étude empirique d'une problème de classification. Ce devoir permettra de me familiriser et comprendre l'implémentation des algorithmes de classification: algorithmes Naives Bayes et Régression logistique.

# 2. Ensemble de donnée


## 2.1 Faire l'importation des librairie nécessaire

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression 
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import precision_score, recall_score
from sklearn.naive_bayes import GaussianNB
import warnings
warnings.filterwarnings("ignore")

## 2.2 Lire l'ensemble de données

J'ai choisi l'ensemble de données sur la qualité des vins à partir de la plateforme Kaggle.

Voici le lien vers la version raw sur de l'ensemble de données sur Github:
https://raw.githubusercontent.com/Hossein1831/KnapSackProblem/main/WineQT.csv

In [2]:
url = 'https://raw.githubusercontent.com/Hossein1831/KnapSackProblem/main/WineQT.csv'
dataset = pd.read_csv(url)

Regardons quelles sont les colonnes de l'ensemble de données: 

In [3]:
dataset.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality', 'Id'],
      dtype='object')

Les colonnes sont: 'fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality' et 'Id'. Sur Kaggle c'est indiqué que les 11 premiers colonnes sont les variables d'entrées et la colonne 'quality' est la variable de sortie.

On peut aussi visualiser les 20 premières rangées de l'ensemble de données.

## 2.3 Étape de prétraitement

Dans cette étape, je vais prétaité l'ensemble de données. La première étape dois toujours être le prétraitement des données. Le prétraitement peut être le nettoyage des données, ordonner les données, enlever des colonnes qui ne sont pas utile, etc.

### 2.3.1 Effacer les colonnes n'ont utile

Dans l'ensemble de données que j'ai choisi il y a la colonne 'Id' qui n'est pas utile, alors je vais l'effacer.

### 2.3.2 Séparation entrée/sortie
 
On peut voir que la colonne 'Id' a était enlever avec succès. Ensuite, j'ai séparé les entrées et les sorties:

In [4]:
inputVariables = dataset.iloc[:, 0:11] # Variables qui contient les colonnes pour les variables d'entrée
outputVariables = dataset.iloc[:, 11:12] # Variables qui contient les colonnes pour la variable de sortie

Nous pouvons visualiser les 10 premières rangées du tableau des entrées:

In [5]:
inputVariables.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides   
0            7.4              0.70         0.00             1.9      0.076  \
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            6.7              0.58         0.08             1.8      0.097   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates   
0                 11.0                  34.0   0.9978  3.51       0.56  \
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 15.0                  65.0   0.9959  3.28       0.54   

   alcohol  
0      9.4  
1      9.8  
2      9.8  
3      9.8  
4      9.4  
5      9.4  
6      9.4  
7     10.0  
8      9.5  
9      9.2

On peut visualiser les 10 premières rangées de la colonne des sorties:

In [6]:
outputVariables.head(10)

quality
0        5
1        5
2        5
3        6
4        5
5        5
6        5
7        7
8        7
9        5

### 2.3.3 Visualisation globale

On peut également utiliser la méthode `describe()` pour avoir une vu des données. Cette méthode nous montre combien de rangées nous avons, les valeurs minimales et maximale de chaque attributs et autre informations.

In [7]:
dataset.describe()

fixed acidity  volatile acidity  citric acid  residual sugar   
count    1143.000000       1143.000000  1143.000000     1143.000000  \
mean        8.311111          0.531339     0.268364        2.532152   
std         1.747595          0.179633     0.196686        1.355917   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.392500     0.090000        1.900000   
50%         7.900000          0.520000     0.250000        2.200000   
75%         9.100000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density   
count  1143.000000          1143.000000           1143.000000  1143.000000  \
mean      0.086933            15.615486             45.914698     0.996730   
std       0.047267            10.250486             32.782130     0.001925   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             21.000000     0.995570   
50%       0.079000            13.000000             37.000000     0.996680   
75%       0.090000            21.000000             61.000000     0.997845   
max       0.611000            68.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality           Id  
count  1143.000000  1143.000000  1143.000000  1143.000000  1143.000000  
mean      3.311015     0.657708    10.442111     5.657043   804.969379  
std       0.156664     0.170399     1.082196     0.805824   463.997116  
min       2.740000     0.330000     8.400000     3.000000     0.000000  
25%       3.205000     0.550000     9.500000     5.000000   411.000000  
50%       3.310000     0.620000    10.200000     6.000000   794.000000  
75%       3.400000     0.730000    11.100000     6.000000  1209.500000  
max       4.010000     2.000000    14.900000     8.000000  1597.000000

### 2.3.4 Nombre de valuer nulle

Ensuite je vais utiliser la méthode `isnull()` et `sum()` pour vérifier le nombre total de valeurs null dans chaque colonnes:

In [8]:
dataset.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
Id                      0
dtype: int64

On peut voir qu'aucune colonne contient des valeurs null. Ceci nous facilite un peu la tâche, car nous aurons pas besoin de faire le traitement de donnée **null**.

### 2.3.5 Séparation de Entrainer et Tester les models

À l'étape **2.3.2** on a séparer les entrées et les sorties. Maintenant on va configurer les variables de *testing* et *training*.

In [9]:
inputVariables_train, inputVariables_test, outputVariables_train, outputVariables_test = train_test_split(inputVariables, outputVariables, random_state=42)

### 2.3.6 Mise à l'echelle

La mise à l'échelle des caractéristiques est une étape importante dans le prétraitement des données pour les algorithmes d'apprentissage automatique, en particulier ceux qui utilisent des techniques d'optimisation comme la régression logistique. Sans mise à l'échelle, des caractéristiques ayant de grandes valeurs pourraient peser inutilement plus dans le modèle. De plus, cela peut ralentir la convergence des algorithmes d'optimisation et rendre le modèle moins efficace. La mise à l'échelle garantit que toutes les caractéristiques contribuent de manière égale à la performance du modèle et accélère la convergence de l'algorithme.

Dans l'ensemble de données que j'ai choisi certains colonnes ont des écarts très grand entre les valeurs et cela peut nuire à l'apprentissage du modèle.

Pour mettre à l'échelle l'ensemble de données je vais utiliser la méthode `StandardScaler()` de **scikit-learn**, cela va nous permettre d'optimiser nos résultats.

In [10]:
standardScaler = StandardScaler()
inputVariables_train_scaled = standardScaler.fit_transform(inputVariables_train)
inputVariables_test_scaled = standardScaler.transform(inputVariables_test)

# 3. Implémentation du model régression logistique

## 3.1 Qu'est-ce que le model régression logistique

La régression logistique est un type d'analyse de régression utilisé pour prédire le résultat d'une variable catégorielle en fonction de variables indépendantes. Dans le contexte de la classification, il est souvent utilisé pour estimer la probabilité qu'un événement se produise ou non. Par exemple, dans l'ensemble de données que j'ai choisi, on pourrait présire si un vin est **bon** ou **mauvais** qualité.

## 3.2 Classifier les qualités en catégorie binaire

Comme le model regréssion logistique est utilisé pour prédire catégoriquement, dans notre cas on doit configurer les critères qui détermine si la qualité d'un vin est **bon** ou **mauvais**.

On va effectuer une transformation de la variable de sortie "quality" en une variable catégorique binaire. Tout d'abord, on divise les valeurs de "quality" en deux catégories en utilisant les seuils définis dans les variables "bins" (2 et 6.5) et les noms de groupe *bad* et *good*. Ensuite, il utilise l'encodeur d'étiquettes pour convertir les catégories en valeurs numériques (0 pour *bad* et 1 pour *good*). Enfin, il affiche les valeurs uniques de la nouvelle variable "quality" et les compte pour chaque catégorie. Les 20 premières lignes de l'ensemble de données modifié sont également affichées.

In [11]:
bins = (2, 6.5, 8)
group_names = ['bad', 'good']
dataset['quality'] = pd.cut(dataset['quality'], bins=bins, labels=group_names)
dataset['quality'].unique()

label_quality = LabelEncoder()
dataset['quality'] = label_quality.fit_transform(dataset['quality'])
dataset['quality'].value_counts()

dataset.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides   
0            7.4              0.70         0.00             1.9      0.076  \
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            6.7              0.58         0.08             1.8      0.097   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates   
0                 11.0                  34.0   0.9978  3.51       0.56  \
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 15.0                  65.0   0.9959  3.28       0.54   

   alcohol  quality  Id  
0      9.4        0   0  
1      9.8        0   1  
2      9.8        0   2  
3      9.8        0   3  
4      9.4        0   4  
5      9.4        0   5  
6      9.4        0   6  
7     10.0        1   7  
8      9.5        1   8  
9      9.2        0  10

On remarque que maintenant au lieu de voir la qualité sur une échelle de 1 à 10, on les visualise simplement par **bon** (1) ou **mauvais** (0).

## 3.3 Entrainer le model de regréssion logistique sur les données

Dans cette étape nous allons entrainer le model de regréssion logistique sur l'ensemble de données que nous avons prétraité précédemment. 

### 3.3.1 Entrainer/tester/évaluer avec paramètre par défaut

D'abord je vais entrainer le model ensuite le stester à l'aide des méthodes de *k-fold cross validation* et ensuite analyser avec la méthode de *macro/micro*

#### Entrainer le model

In [12]:
logisticRegression_default = LogisticRegression(max_iter=100)
logisticRegression_default.fit(inputVariables_train_scaled, outputVariables_train.values.ravel()) #Entraîner le modèle avec les données d'entraînement
Predictions = logisticRegression_default.predict(inputVariables_test_scaled) #Prédire les données de test

En entrainant le model avec les paramètres de test_size et random_state configurés à leur valeurs par défaut, j'ai atteint une précision de 58.74%. La précision des prédictions est effectivement vraiment bas. Nous devons trouvons une façon d'améliorer les prédictions. Pour faire cela, je vais utiliser la méthode de *k-fold cross validation*. *K-fold cross validation* est une technique utilisée pour évaluer la performance d'un modèle d'apprentissage automatique en divisant l'ensemble de données en k sous-ensembles (*folds*). Le modèle est formé sur k−1 de ces *folds* et testé sur le *fold* restant, en répétant ce processus `k` fois avec différents *folds* utilisés comme ensemble de test à chaque fois. La moyenne de la métrique de performance sur les `k` tests est utilisée comme mesure globale de la performance du modèle.

#### Tester (4-fold cross validation & micro/macro)

Je vais tester le model que je viens d'entrainer. Je vais d'abord commencer avec en testant avec la technique de *k-fold cross validation* et ensuite à l'aide des moyennes **macro** et **micro**.

##### 4-fold cross validation :

In [13]:
#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(logisticRegression_default, inputVariables, outputVariables, cv=4) 

#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 

#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle sur les 4 folds est: %.2f%%" % (meanCrossValScore * 100)) 

La précision moyenne du modèle sur les 4 folds est: 56.78%


##### Micro & macro :

In [14]:
micro_precision = precision_score(outputVariables_test, Predictions, average='micro')
macro_precision = precision_score(outputVariables_test, Predictions, average='macro')

micro_recall = recall_score(outputVariables_test, Predictions, average='micro')
macro_recall = recall_score(outputVariables_test, Predictions, average='macro')

print(f"Précision de moyenne micro: {micro_precision}")
print(f"Précision de moyenne macro: {macro_precision}")
print(f"Rapelle de moyenne micro:   {micro_recall}")
print(f"Rapelle de moyenne macro:   {macro_recall}")

Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723


### 3.3.2 Entrainer/tester/évaluer avec paramètre de *solver* différent

Cette fois je vais refaire les même étapes mais avec des paramètre de régression linéaire différente. Je vais seulement changer le paramètre de `solver` différent. La valeur par défault de cette paramètre est `'lbfgs'`. Je vais configurer une fois la valeur à `'liblinear'` et une autre fois à `'newtong-cg'` et ensuite comparer les résultats avec les tests de *k-folding* et *micro/macro*.

#### Entrainer

In [15]:
logisticRegression_change1 = LogisticRegression(max_iter=100, solver='liblinear') # Valeur de Solver configurer à liblinear
logisticRegression_change1.fit(inputVariables_train_scaled, outputVariables_train.values.ravel()) #Entraîner le modèle avec les données d'entraînement
Predictions_change1 = logisticRegression_change1.predict(inputVariables_test_scaled) #Prédire les données de test

logisticRegression_change2 = LogisticRegression(max_iter=100, solver='newton-cg') # Valeur de Solver configurer à newton-cg
logisticRegression_change2.fit(inputVariables_train_scaled, outputVariables_train.values.ravel()) #Entraîner le modèle avec les données d'entraînement
Predictions_change2 = logisticRegression_change2.predict(inputVariables_test_scaled) #Prédire les données de test

#### Tester (4-fold cross validation & micro/macro)

Je vais tester le model que je viens d'entrainer. Je vais d'abord commencer avec en testant avec la technique de *k-fold cross validation* et ensuite à l'aide des moyennes **macro** et **micro**.

##### 4-fold cross validation :

In [16]:
'''change 1'''
#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(logisticRegression_change1, inputVariables, outputVariables, cv=4) 
#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 
#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle (liblinear) sur les 4 folds est: %.2f%%" % (meanCrossValScore * 100)) 

'''change 2'''
#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(logisticRegression_change2, inputVariables, outputVariables, cv=4) 
#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 
#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle (newton-cg) sur les 4-folds est: %.2f%%" % (meanCrossValScore * 100)) 

La précision moyenne du modèle (liblinear) sur les 4 folds est: 58.09%
La précision moyenne du modèle (newton-cg) sur les 4-folds est: 58.88%


##### Micro & macro (solver='liblinear'):

In [17]:
micro_precision = precision_score(outputVariables_test, Predictions_change1, average='micro')
macro_precision = precision_score(outputVariables_test, Predictions_change1, average='macro')

micro_recall = recall_score(outputVariables_test, Predictions_change1, average='micro')
macro_recall = recall_score(outputVariables_test, Predictions_change1, average='macro')

print(f"Précision de moyenne micro: {micro_precision}")
print(f"Précision de moyenne macro: {macro_precision}")
print(f"Rapelle de moyenne micro:   {micro_recall}")
print(f"Rapelle de moyenne macro:   {macro_recall}")

Précision de moyenne micro: 0.6328671328671329
Précision de moyenne macro: 0.3821480831138425
Rapelle de moyenne micro:   0.6328671328671329
Rapelle de moyenne macro:   0.34569761789013126


In [18]:
##### Micro & macro (solver='newton-cg'):

In [19]:
micro_precision = precision_score(outputVariables_test, Predictions_change2, average='micro')
macro_precision = precision_score(outputVariables_test, Predictions_change2, average='macro')

micro_recall = recall_score(outputVariables_test, Predictions_change2, average='micro')
macro_recall = recall_score(outputVariables_test, Predictions_change2, average='macro')

print(f"Précision de moyenne micro: {micro_precision}")
print(f"Précision de moyenne macro: {macro_precision}")
print(f"Rapelle de moyenne micro:   {micro_recall}")
print(f"Rapelle de moyenne macro:   {macro_recall}")

Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723


### 3.3.3 Entrainer/tester/évaluer avec paramètre de *tol* différent

Cette fois je vais refaire les même étapes mais avec des paramètre de régression linéaire différente. Je vais seulement changer le paramètre de `tol` (tolérence) différent. La valeur par défault de cette paramètre est `tol=1e-4`. Je vais configurer une fois la valeur à `tol=1e-1` et une autre fois à `tol=1e-6` et ensuite comparer les résultats avec les tests de *k-folding* et *micro/macro*.

In [20]:
logisticRegression_change3 = LogisticRegression(max_iter=100, tol=0.1) # Valeur de Solver configurer à liblinear
logisticRegression_change3.fit(inputVariables_train_scaled, outputVariables_train.values.ravel()) #Entraîner le modèle avec les données d'entraînement
Predictions_change3 = logisticRegression_change3.predict(inputVariables_test_scaled) #Prédire les données de test

logisticRegression_change4 = LogisticRegression(max_iter=100, tol=0.000001) # Valeur de Solver configurer à newton-cg
logisticRegression_change4.fit(inputVariables_train_scaled, outputVariables_train.values.ravel()) #Entraîner le modèle avec les données d'entraînement
Predictions_change4 = logisticRegression_change4.predict(inputVariables_test_scaled) #Prédire les données de test

#### Tester (4-fold cross validation & micro/macro)

Je vais tester le model que je viens d'entrainer. Je vais d'abord commencer avec en testant avec la technique de *k-fold cross validation* et ensuite à l'aide des moyennes **macro** et **micro**.

##### 4-fold cross validation :

In [21]:
'''change 3'''
#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(logisticRegression_change3, inputVariables, outputVariables, cv=4) 
#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 
#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle (tol=0.1) sur les 4 folds est: %.2f%%" % (meanCrossValScore * 100)) 

'''change 4'''
#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(logisticRegression_change4, inputVariables, outputVariables, cv=4) 
#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 
#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle (tol=0.000001) sur les 4 folds est: %.2f%%" % (meanCrossValScore * 100)) 

La précision moyenne du modèle (tol=0.1) sur les 4 folds est: 56.78%
La précision moyenne du modèle (tol=0.000001) sur les 4 folds est: 56.78%


##### Micro & macro tol=1e-1 :

In [22]:
micro_precision = precision_score(outputVariables_test, Predictions_change3, average='micro')
macro_precision = precision_score(outputVariables_test, Predictions_change3, average='macro')

micro_recall = recall_score(outputVariables_test, Predictions_change3, average='micro')
macro_recall = recall_score(outputVariables_test, Predictions_change3, average='macro')

print(f"Précision de moyenne micro: {micro_precision}")
print(f"Précision de moyenne macro: {macro_precision}")
print(f"Rapelle de moyenne micro:   {micro_recall}")
print(f"Rapelle de moyenne macro:   {macro_recall}")

Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723


##### Micro & Macro tol=1e-6 :

In [23]:
micro_precision = precision_score(outputVariables_test, Predictions_change4, average='micro')
macro_precision = precision_score(outputVariables_test, Predictions_change4, average='macro')

micro_recall = recall_score(outputVariables_test, Predictions_change4, average='micro')
macro_recall = recall_score(outputVariables_test, Predictions_change4, average='macro')

print(f"Précision de moyenne micro: {micro_precision}")
print(f"Précision de moyenne macro: {macro_precision}")
print(f"Rapelle de moyenne micro:   {micro_recall}")
print(f"Rapelle de moyenne macro:   {macro_recall}")

Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723


# 4. Implémentation du model Naïve Bayes Gaussian

## 4.1 Qu'est-ce que le model Naïve Bayes Gaussian

J'ai choisi le modèle Naïve Bayes Gaussien qui est une variante de l'algorithme Naïve Bayes utilisée pour les caractéristiques continues. Ce modèle suppose que les valeurs de chaque caractéristique suivent une distribution gaussienne (ou normale). L'algorithme calcule la probabilité de chaque classe pour une observation donnée en utilisant la formule de Bayes et en supposant que les caractéristiques sont indépendantes les unes des autres. Ensuite, il attribue l'observation à la classe ayant la plus grande probabilité. Ce modèle est particulièrement utile pour les tâches de classification rapide et simple.

## 4.2 Entrainer/tester/évaluer

D'abord je vais entrainer le model ensuite le stester à l'aide des méthodes de *k-fold cross validation* et ensuite analyser avec la méthode de *macro/micro*

### 4.2.1 Entrainer/tester/évaluer le model avec paramètre par défault

Je vais débuter en entrainant, testant et évaluant le premier modèle qui a des paramètres configurer à leurs valeur par défault.

#### Entrainer

In [24]:
gaussian_default = GaussianNB()
prediction_gaussian_default = gaussian_default.fit(inputVariables_train, outputVariables_train).predict(inputVariables_test)
print("Default Naive Bayes Accuracy : ", accuracy_score(outputVariables_test, prediction_gaussian_default, normalize = True))

Default Naive Bayes Accuracy :  0.583916083916084


#### Tester

Dans cette étape je vais tester le modèle que je viens d'entrainer en utilisant la technique de *k-fold cross validation*

##### 4-fold cross valiadtion

In [25]:
#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(gaussian_default, inputVariables, outputVariables, cv=4)

#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 

#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle sur les 4 folds est: %.2f%%" % (meanCrossValScore * 100))

La précision moyenne du modèle sur les 4 folds est: 52.59%


#### Évaluer

Je vais utiliser la méthode micro/macro des moyenne de rapelle et de précision.

In [26]:
micro_precision = precision_score(outputVariables_test, prediction_gaussian_default, average='micro')
macro_precision = precision_score(outputVariables_test, prediction_gaussian_default, average='macro')

micro_recall = recall_score(outputVariables_test, prediction_gaussian_default, average='micro')
macro_recall = recall_score(outputVariables_test, prediction_gaussian_default, average='macro')

print(f"Micro-Average Precision: {micro_precision}")
print(f"Macro-Average Precision: {macro_precision}")
print(f"Micro-Average Recall   : {micro_recall}")
print(f"Macro-Average Recall   : {macro_recall}")

Micro-Average Precision: 0.583916083916084
Macro-Average Precision: 0.30468913580349627
Micro-Average Recall   : 0.583916083916084
Macro-Average Recall   : 0.30051855452924975


### 4.2.2 Entrainer/tester/évaluer avec paramètre modifiée

Maintenant je vais modifier une des paramètres de la méthode `gaussionNB()`. Je vais modifier le paramètre `var_smoothing`. Par défault cette paramètre est configurer à `var_smoothing=1e-9`. Je vais le configurer une fois à `var_smoothing=1e-1` et une autre fois à `var_smoothing=1e-5`.

#### Entrainer

In [27]:
'''change 1'''
gaussian_change1 = GaussianNB(var_smoothing=1e-1)
prediction_gaussian_change1 = gaussian_change1.fit(inputVariables_train, outputVariables_train).predict(inputVariables_test)
print("Naive Bayes Accuracy : ", accuracy_score(outputVariables_test, prediction_gaussian_change1, normalize = True))

'''change 2'''
gaussian_change2 = GaussianNB(var_smoothing=1e-5)
prediction_gaussian_change2 = gaussian_change2.fit(inputVariables_train, outputVariables_train).predict(inputVariables_test)
print("Naive Bayes Accuracy : ", accuracy_score(outputVariables_test, prediction_gaussian_change2, normalize = True))

Naive Bayes Accuracy :  0.5174825174825175
Naive Bayes Accuracy :  0.6153846153846154


#### Tester

Dans cette étape je vais tester le modèle que je viens d'entrainer en utilisant la technique de k-fold cross validation.

##### 4-fold cross valiadtion

In [28]:
#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(gaussian_change1, inputVariables, outputVariables, cv=4)
#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 
#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle (1e-1) sur les 4 folds est: %.2f%%" % (meanCrossValScore * 100))

#Calculer la précision du modèle avec la validation croisée à 4-fold
crossValidation = cross_val_score(gaussian_change2, inputVariables, outputVariables, cv=4)
#Calculer la précision moyenne du modèle sur les 4 folds
meanCrossValScore = np.mean(crossValidation) 
#Afficher la précision moyenne du modèle sur les 4 folds en pourcentage
print("La précision moyenne du modèle (1e-5) sur les 4 folds est: %.2f%%" % (meanCrossValScore * 100))

La précision moyenne du modèle (1e-1) sur les 4 folds est: 48.20%
La précision moyenne du modèle (1e-5) sur les 4 folds est: 54.77%


#### Évaluer 

Je vais utiliser la méthode de micro/macro pour la analyser les moyenne de rapelle et précision.

##### Évaluation pour `gaussian_change1`

In [29]:
micro_precision = precision_score(outputVariables_test, prediction_gaussian_change1, average='micro')
macro_precision = precision_score(outputVariables_test, prediction_gaussian_change1, average='macro')

micro_recall = recall_score(outputVariables_test, prediction_gaussian_change1, average='micro')
macro_recall = recall_score(outputVariables_test, prediction_gaussian_change1, average='macro')

print(f"Micro-Average Precision: {micro_precision}")
print(f"Macro-Average Precision: {macro_precision}")
print(f"Micro-Average Recall   : {micro_recall}")
print(f"Macro-Average Recall   : {macro_recall}")

Micro-Average Precision: 0.5174825174825175
Macro-Average Precision: 0.22632045253186125
Micro-Average Recall   : 0.5174825174825175
Macro-Average Recall   : 0.24462809917355371


##### Évaluation pour `gaussian_change2`

In [30]:
micro_precision = precision_score(outputVariables_test, prediction_gaussian_change2, average='micro')
macro_precision = precision_score(outputVariables_test, prediction_gaussian_change2, average='macro')

micro_recall = recall_score(outputVariables_test, prediction_gaussian_change2, average='micro')
macro_recall = recall_score(outputVariables_test, prediction_gaussian_change2, average='macro')

print(f"Micro-Average Precision: {micro_precision}")
print(f"Macro-Average Precision: {macro_precision}")
print(f"Micro-Average Recall   : {micro_recall}")
print(f"Macro-Average Recall   : {macro_recall}")

Micro-Average Precision: 0.6153846153846154
Macro-Average Precision: 0.30498800959232614
Micro-Average Recall   : 0.6153846153846154
Macro-Average Recall   : 0.31291524874412574


# 5. Analyse et comparaison des résulats obtenus selon les modèles entrainés

Dans cette étape je vais analyser les résultats obtenus et de comprendre leurs signification, et ensuite les comparer avec d'autre résultats où certains paramètres étaient modifées.

J'ai entrainer deux modèles sur un même ensemble de données. J'ai choisi les modèles de **regréssion linéaire** et **Naïves Bayes Gaussian**. Pour le premier entrainement de chaque modèle j'ai utiliser les modèles configurés à leurs paramètres par défault. J'ai ensuite effectué un testing en utilisant la technique de *k-fold cross-validation* avec `k = 4` et ensuite j'ai évaluer les modèles en utilisant les moyennes *micro* et *macro* de rapelle et précision. Par la suite, j'ai effectué les même étapes pour des modèles avec des paramètres modifiées pour expérimenter et de mieux comprendre l'impact des paramètres et des modèles.

## 5.1 Analyse du modèle de regréssion linéaire avec les paramètres par défault

### Entrainement

J'ai entrainer le modèle avec les paramètres configurées à leurs valeurs par défault.

### Tester

J'ai utilisé la technique de *k-fold cross-validation* avec `k = 4`. J'ai obtenu le résultat suivant: 
```
La précision moyenne du modèle sur les 4 folds est: 56.78%.
```
Cela représente la capacité du modèle à bien généraliser sur de nouvelles données. Par contre, une valeur de `56.78%` signifie que le modèle est plutôt mauvais. Plusieurs facteurs peuvent avoir causer l'obtention d'une valeur si bas. Notamment, la configuration des paramètres du modèle à leurs valeurs par défaut. Cela peut indiquer que configurer certains paramètres à leurs valeurs par défault ne soient pas adaptés pour ce problème. 

### Évaluation 

J'ai utilisé la technique des moyennes *micro* et *macro* de rapelle et précision. Et j'ai obtenu les résultats suivantes:
```
Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723
```
**Précision de moyenne micro** et **Rapelle de moyenne micro** sont identiques. Cela peut arriver quand la distribution des classes est relativement uniforme, ou quand les erreurs dans les différentes classes se compensent mutuellement. Les valeurs faibles de **précision de moyenne macro** et **rapelle de moyenne macro** suggèrent que le modèle a du mal à classer certaines classes, ce qui peut arriver en présence de classes déséquilibrées ou moins représentées. La **précision de moyenne micro** est plus élevée que la **précision de moyenne macro** ca veut dire que le modèle est probablement bon pour les classes majoritaires mais pas aussi bon pour les classes minoritaires.

## 5.2 Analyse du modèle de regréssion linéaire avec paramètre de `solver` différent

### Entrainement

Cette fois j'ai modifié le paramètre `solver`. J'y ai donné deux valeurs différente que celui par défaut. Par défaut, la valeur est `solver='lbfgs'` et j'ai implémenté une fois avec `solver='liblinear'` et une fois avec `solver='newton-cg'`.

### Tester

J'ai utilisé la technique de *k-fold cross-validation* avec `k = 4` pour chaque implémentation. J'ai obtenu les résultats suivants:
```
La précision moyenne du modèle (liblinear) sur les 4 folds est: 58.09% 
La précision moyenne du modèle (newton-cg) sur les 4 folds est: 58.88%
``` 
Dans mes tests, les `solver` `liblinear` et `newton-cg` ont produit des résultats de précision assez similaires en validation croisée. On peut dire que dans ce cas, le `solver='newton-cg'` a mieux performer que `solver='liblinear'`. Même si les valeurs obtenus sont plus élevé que le résultat obtenu avec le paramètre par défaut `56.78%`, il n'y a pas une si grande différence. Ce qui suggère que le choix du `solver` n'a pas un impact significatif sur cet ensemble de données.

### Évaluation

J'ai utilisé la technique des moyennes *micro* et *macro* de rapelle et précision pour chacun des implémentations. Et j'ai obtenu les résultats suivantes:
```
**Qaund solver='liblinear'**
Précision de moyenne micro: 0.6328671328671329
Précision de moyenne macro: 0.3821480831138425
Rapelle de moyenne micro:   0.6328671328671329
Rapelle de moyenne macro:   0.34569761789013126

**Quand solver='newton-cg'**
Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723
``` 
Les mesures de précision et de rappel *micro* sont également similaires, indiquant que les deux solveurs sont presque également efficaces sur l'ensemble des échantillons. Cependant, les mesures *macro* varient, montrant des différences dans la manière dont les solveurs traitent les différentes classes. La similarité des résultats entre les solveurs spécifiés et le `solver` par défaut indique que le problème est assez bien conditionné pour que différents algorithmes d'optimisation arrivent à des solutions comparables.

## 5.3 Analyse du modèle de regréssion linéaire avec paramètre de `tol` différent

### Entrainer

Cette fois j'ai modifié le paramètre `tol`. J'y ai donné deux valeurs différente que celui par défaut. Par défaut, la valeur est `tol=1e-4` et j'ai implémenté une fois avec `tol=1e-1` et une fois avec `tol=1e-6`.

### Tester

J'ai utilisé la technique de *k-fold cross-validation* avec `k = 4` pour chaque implémentation. J'ai obtenu les résultats suivants:
```
La précision moyenne du modèle (tol=0.1) sur les 4 folds est: 56.78%
La précision moyenne du modèle (tol=0.000001) sur les 4 folds est: 56.78%
```
Les deux modèles de régression logistique avec différentes valeurs de tolérance (tol=0.1 et tol=0.000001) ont produit des résultats identiques en termes de précision en validation croisée (56,78%).

### Évaluer

J'ai utilisé la technique des moyennes *micro* et *macro* de rapelle et précision pour chacun des implémentations. Et j'ai obtenu les résultats suivantes:
```
**Quand tol=1e-1** :
Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723

**Quand tol=1e-6** :
Précision de moyenne micro: 0.6363636363636364
Précision de moyenne macro: 0.30845797720797724
Rapelle de moyenne micro:   0.6363636363636364
Rapelle de moyenne macro:   0.3035569599740723
```
Les deux modèles de régression logistique avec différentes valeurs de tolérance (tol=0.1 et tol=0.000001) ont produit des résultats identiques en termes de précision en validation croisée (56,78%) et en précision et rappel micro et macro. Cela suggère que la tolérance, qui est le critère d'arrêt pour le solveur, n'a pas eu d'impact significatif sur les performances du modèle pour cet ensemble de données. Le fait que ces résultats soient également identiques à ceux obtenus avec la tolérance par défaut indique que la valeur par défaut de la tolérance dans scikit-learn est probablement déjà bien adaptée pour ce problème. La précision en validation croisée de 56,78% donne une idée de la performance généralisée du modèle, tandis que les mesures de précision et de rappel micro et macro reflètent la performance du modèle sur les différentes classes.

## 5.4 Analyse du modèle de Naïves Bayes Gaussian avec paramètre par défault

### Entrainer

J'ai entrainer le modèle avec les paramètres configurées à leurs valeurs par défault.

### Tester

J'ai utilisé la technique de *k-fold cross-validation* avec `k = 4`. J'ai obtenu le résultat suivant: 
```
La précision moyenne du modèle sur les 4 folds est: 52.59%.
```
La précision moyenne de 52.59% obtenue par la validation croisée à 4 folds indique que le modèle Naïve Bayes Gaussien est moins performant que les modèles de régression logistique précédents. Cela pourrait être dû à la distribution des données ou à la manière dont Naïve Bayes traite les caractéristiques.

### Évaluer

J'ai utilisé la technique des moyennes *micro* et *macro* de rapelle et précision. Et j'ai obtenu les résultats suivantes:
```
Micro-Average Precision: 0.583916083916084
Macro-Average Precision: 0.30468913580349627
Micro-Average Recall   : 0.583916083916084
Macro-Average Recall   : 0.30051855452924975
```
La précision et le rappel "micro-moyens" sont identiques (environ 58.39%), ce qui signifie que le modèle a une performance générale assez cohérente sur tous les échantillons. Le fait que les mesures "micro" et "macro" ne soient pas les mêmes indique également que la distribution des classes dans les données n'est probablement pas uniforme.

## 5.5 Analyse du modèle de Naïves Bayes Gaussian avec paramètre de `var_smoothing` différent

### Entrainer

Cette fois j'ai modifié le paramètre `var_smoothing`. J'y ai donné deux valeurs différente que celui par défaut. Par défaut, la valeur est `var_smoothing=1e-9` et j'ai implémenté une fois avec `var_smoothing=1e-1` et une fois avec `var_smoothing=1e-5`.

### Tester

J'ai utilisé la technique de *k-fold cross-validation* avec `k = 4` sur les deux modèles. J'ai obtenu le résultat suivant: 
```
La précision moyenne du modèle (1e-1) sur les 4 folds est: 48.20%
La précision moyenne du modèle (1e-5) sur les 4 folds est: 54.77%
```
La précision moyenne obtenue en validation croisée à 4-folds varie en fonction de la valeur de `var_smoothing`. Avec `1e-1`, elle est de `48.20%`, et avec `1e-5`, elle monte à `54.77%`. Cela suggère que la valeur de `var_smoothing=1e-5` est plus adaptée à vos données. 
Parce que ce dernier a une meilleure *cross-validation* que la valeur par défaut.

### Évaluer

J'ai utilisé la technique des moyennes *micro* et *macro* de rapelle et précision pour chacun des implémentations. Et j'ai obtenu les résultats suivantes:

**Quand `var_smoothing = 1e-1`**:
```
Micro-Average Precision: 0.5174825174825175
Macro-Average Precision: 0.22632045253186125
Micro-Average Recall   : 0.5174825174825175
Macro-Average Recall   : 0.24462809917355371
```
**Quand `var_smoothing = 1e-5`**:
```
Micro-Average Precision: 0.6153846153846154
Macro-Average Precision: 0.30498800959232614
Micro-Average Recall   : 0.6153846153846154
Macro-Average Recall   : 0.31291524874412574
```
La précision et le rappel micro-moyens sont les mêmes dans chaque cas, indiquant une performance globale cohérente sur l'ensemble du jeu de données. Cependant, ces valeurs augmentent lorsque `var_smoothing` passe de `1e-1` à `1e-5`. Les précisions et rappels **macro-moyens** sont également plus faibles que les **micro-moyens**, mais ils s'améliorent également lorsque `var_smoothing` est réduit de `1e-1` à `1e-5`.

# 6. Conclusion

En conclusion, selon le *4-fold cross validation* les deux modèles sont proches. Selon mes données et les paramètres que j'ai modifié, le modèle de regréssion linéaire avec le paramètre de `solver` configurer à `newton-cg` a obtenu le meilleur moyenne de *cross-validation*. Ceci est pouvais être différent si j'avais changer d'autres paramètres. Par contre, en analysant les précisions et rapelles de **macro** et **micro** j'ai remarqué que tous les diférent configuration du modèle de **régression linéaire** ont mieux performé que les modèles de **Naïves Bayes Gaussian**. Seul le modèle de **Naïves Bayes Gaussian** avec le paramètre de `var_smoothing = 1e-5` a performé au même niveau que les modèles de **rgréssion linéaire**. J'ai obtenu des valeurs plus grands dans les moyennes **micro** et des valeurs plus petite dans les moyennes **macro**, cela démontre que mes modèles performent mieux sur les classes majoritaire mais pas nécessairement sur les classes minoritaire. Les ***classes majoritaires** sont celles qui ont le plus grand nombre d'échantillons dans l'ensemble de données, dans mon cas ce sont les données classé comme **mauvais**. Les **classes minoritaires** sont celles qui ont le plus petit nombre d'échantillons dans l'ensemble de données, dans mon cas les données classées comme **bon**. Ceci peut être un grand problème parce que le système va être biasé à classer les vins comme mauvais, même si les attributs du vin sont orientés plus vers un bon vin.

# 7. Référence

Youtube -> Simplilearn: Scikit-Learn Tutorial | Machine Learning With Scikit-Learn | Sklearn | Python Tutorial | Simplilearn

Link -> https://youtu.be/0Lt9w-BxKFQ?si=da-s3zA8qmW9e8bB

--------------------------------

Youtube -> Bevan Smith 2: IML28: Logistic regression (part 1): Introduction

Link -> https://youtu.be/vOl_pKenYTY?si=o8hkqGqUDrxQGSQ_

--------------------------------

Youtube -> The AI & DS Channel: Wine quality prediction using logistic regression | Machine Learning Project 3

Link -> https://youtu.be/UkzV1e4tRyk?si=Xzb50QQTOJtPPLYX

--------------------------------

Youtube -> Bevan Smith 2: ML9: How to train and test a simple model using Scikit-learn in Jupyter Notebook (part 1)

Link -> https://youtu.be/bXB3xYh9LeA?si=gF_IZIonZmE4Hjxg

--------------------------------

Youtube -> Bevan Smith 2: ML9: How to train and test a simple model using Scikit-learn in Jupyter Notebook (part 2)

Link -> https://youtu.be/MQPZ3arjLLg?si=ehebA5ZBbwnJSRUF

--------------------------------

Site web -> sklearn.linear_model.LogisticRegression

Link -> https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression

--------------------------------

Site web -> scikit-learn: 1.9. Naive Bayes

Link -> https://scikit-learn.org/stable/modules/naive_bayes.html

--------------------------------

Site web -> sklearn.naive_bayes.GaussianNB

Link -> https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html

--------------------------------

Site web -> Kaggle: Wine Quality Dataset

Link -> https://www.kaggle.com/datasets/yasserh/wine-quality-dataset
